Source/swiftlint/Helpers/LintableFilesVisitor.swift

```diff
@@ -93,25 +109,69 @@ struct LintableFilesVisitor {
         switch self.mode {
         case .lint:
             return Linter(file: file, configuration: configuration, cache: cache)
-        case let .analyze(allCompilerInvocations):
-            let compilerArguments = file.path.flatMap {
-                CompilerArgumentsExtractor.compilerArgumentsForFile($0, compilerInvocations: allCompilerInvocations)
-            } ?? []
+        case let .analyze(compilerInvocations):
+            let compilerArguments = compilerInvocations.arguments(forFile: file.path)
             return Linter(file: file, configuration: configuration, compilerArguments: compilerArguments)
         }
     }
 
-    private static func compilerLogContents(logPath: String) -> String? {
-        if logPath.isEmpty {
-            return nil
+    private static func loadCompilerInvocations(_ options: LintOrAnalyzeOptions)
+        -> Result<CompilerInvocations, CommandantError<()>> {
+        if !options.compilerLogPath.isEmpty {
+            let path = options.compilerLogPath
+            guard let compilerInvocations = self.loadLogCompilerInvocations(path) else {
+                return .failure(
+                    .usageError(description: "Could not read compiler log at path: '\(path)'")
+                )
+            }
+
+            return .success(.buildLog(compilerInvocations: compilerInvocations))
+        } else if !options.compileCommands.isEmpty {
+            let path = options.compileCommands
+            guard let compileCommands = self.loadCompileCommands(path) else {
+                return .failure(
+                    .usageError(description: "Could not read compilation database at path: '\(path)'")
+                )
+            }
+
+            return .success(.compilationDatabase(compileCommands: compileCommands))
         }
 
-        if let data = FileManager.default.contents(atPath: logPath),
+        return .failure(.usageError(description: "Could not read compiler invocations"))
+    }
+
+    private static func loadLogCompilerInvocations(_ path: String) -> [String]? {
+        if let data = FileManager.default.contents(atPath: path),
             let logContents = String(data: data, encoding: .utf8) {
-            return logContents
+            if logContents.isEmpty {
+                return nil
+            }
+
+            return CompilerArgumentsExtractor.allCompilerInvocations(compilerLogs: logContents)
         }
 
-        print("couldn't read log file at path '\(logPath)'")
         return nil
     }
+
+    private static func loadCompileCommands(_ path: String) -> [String: [String]]? {
+        guard let jsonContents = FileManager.default.contents(atPath: path),
+            let object = try? JSONSerialization.jsonObject(with: jsonContents),
+            let compileDB = object as? [[String: Any]] else {
+            return nil
+        }
+
+        // Convert the compilation database to a dictionary, with source files as keys and compiler arguments as values.
```

Comment:
A note about this dictionary. It could use quite a lot of redundant memory for large modules. For each file in a module, there will be a single compiler invocation (listing each file in the module). For example:

```
swiftc --yada --yada File1.swift File2.swift ... FileN.swift 
```

This command line will be copied N times, one for each path in the compilation database. In theory, the same array could be shared for each file in the module. Array deduping (or string interning) would reduce the memory here.

Just wanted to make a note of this.